### ISSUE TITLE
v1.9.1 unable to remove network due to active endpoints

### ISSUE THREAD

**Issue by Toady00** [docker/docker#19261]
This is the same symptoms of #17555 but I am experiencing it with 1.9.1 and since #17558 was suppose to fix it, I'm assuming this should be a new issue.

I have a Jenkins server running tests for a rails app. The build script sets up all the external dependencies in containers, runs the tests, and then removes all the containers. It was running on a t2.micro (single core) on AWS with no problems. I scaled it up to a t2.medium (2 cores) yesterday and the first time I ran the build, it failed when trying to remove the network, due to still having active endpoints.

The build script is basically cleaning up anything left over from older builds (in case of a failure), and then starts fresh. Here is the relevant portion of the build script:

```bash
docker stop ${db} || echo 'db already stopped'
docker rm ${db} || echo 'db already cleaned up'
docker network rm ${net} || echo 'network already cleaned up'

docker network create --driver bridge ${net}
docker run -d --net ${net} --name ${db} postgres
```

Removing the network doesn't actually fail the build since the `||` is there, but it is erring with `active endpoints`. It actually fails when trying to create the network, since it already exists.

I can see there are no containers connected to the network, but I can't remove the network. As a work around, I don't mind using a `sleep` between creating the network and running the container, but I'd love to know if there is a way to force remove the zombie network.

```bash
$ sudo docker network inspect demo_net
[
    {
        "Name": "demo_net",
        "Id": "ec089055c30c327935772574961f1b25dd2701c4d27ea1e34a70960595bd5f1f",
        "Scope": "local",
        "Driver": "bridge",
        "IPAM": {
            "Driver": "default",
            "Config": [
                {}
            ]
        },
        "Containers": {},
        "Options": {}
    }
]
$ sudo docker network rm demo_net
Error response from daemon: network demo_net has active endpoints
```

Here's all the basic docker stuff:

```bash
$ sudo docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

$ sudo docker info
Containers: 8
Images: 133
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 149
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-48-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 2
Total Memory: 3.859 GiB
Name: jenkins
ID: UQJY:AGBB:HSJL:6XSW:4KBD:EVAB:W6A3:X7XN:RJRT:JDLO:MJBL:D3ZS
WARNING: No swap limit support

$ uname -a
Linux jenkins 3.13.0-48-generic #80-Ubuntu SMP Thu Mar 12 11:16:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

t2.medium AWS instance running an official Ubuntu 14.04 AMI.

**Comment by thaJeztah** [comment]
I know many improvement were made in libnetwork for docker 1.10 (rc1 has been released today for testing; https://github.com/docker/docker/releases/tag/v1.10.0-rc1)

ping @aboch do you know if this should be fixed in 1.10?

**Comment by lueliu** [comment]
I had the same problem.when I use "docker network inspect" ,show ""Containers": {}". 
I have configured etcd inside The docker. I found it (endpoints) in etcd, and the network ID I want to remove.It is show that "network ID: count:14" in etcd. I change the 14 to 0. And then I can  delete the network successfully. 

my version: docker 1.9.0  and network plugin:weave 1.4.1

**Comment by ggtools** [comment]
I had the same problem after upgrading to 1.10 and I solved it the same way with Consul:

    curl -XPUT http://127.0.0.1:8500/v1/kv/docker/network/v1.0/endpoint_count/<network id>/ -d '{"Count":0}'

**Comment by adaptivdesign** [comment]
1.10 Same problem here. Changing the network count to 0 allowed me to remove my overlay network.

**Comment by npcode** [comment]
v1.10.3 Same problem here in bridge network.

**Comment by npcode** [comment]
I solved my problem by restarting docker daemon.

**Comment by lavalamp-** [comment]
+1 Would love to know when this will be addressed.

**Comment by ChrisMcKenzie** [comment]
I am experiencing this on docker v1.10.2 with overlay

**Comment by mavenugo** [comment]
We fixed a few issues for the bridge driver in 1.11 and we would encourage folks to try the 1.11-RC.

But if the issue is still seen in 1.11-RC for either bridge or overlay driver, pls execute `docker network inspect {network-name}` where there can be containers or stale endpoint still left over. One can disconnect the container/endpoint using `docker network disconnect -f {network-name} {endpoint-name}`. Once all the containers / endpoints are disconnected, the network can be removed.

**Comment by brettdh** [comment]
Bit of an extreme, but I encountered this with docker 1.11.0 after a power outage; the containers running before the power loss were still listed as active endpoints, even after I removed them.

**Comment by deitch** [comment]
And `network disconnect` doesn't work, because it cannot find the endpoints!

````
[root@target1 ~]# docker network disconnect -f mynet sad_sinoussi
Error response from daemon: endpoint sad_sinoussi not found
````

Any way to just clear this out of etcd in the meantime?

**Comment by juliangamble** [comment]
We had this issue too. 

Eventually we rolled our docker-compose.yml file back from v2 syntax to v1 syntax and the issue went away. 

Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:20:08 2016
 OS/Arch:      windows/amd64

Server:
 Version:      swarm/1.2.4
 API version:  1.22
 Go version:   go1.5.4
 Git commit:   5d5f7f0
 Built:        Thu Jul 28 19:52:54 UTC 2016
 OS/Arch:      linux/amd64

**Comment by randunel** [comment]
It happened in 1.9.x, it happened in 1.10.x, it happened in 1.11.x and now it's happening to us in 1.12.0.

```
$ docker network rm ub_london
Error response from daemon: Error response from daemon: network ub_london has active endpoints
$ docker network inspect ub_london
[
    {
        "Name": "ub_london",
        "Id": "6ca5f4125ed6093b259aa534dd9a175181c794b0c810e544b7fda5b05bec3f20",
        "Scope": "global",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.0.0/24",
                    "Gateway": "10.0.0.1/24"
                }
            ]
        },
        "Internal": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
$ docker network rm ub_london
Error response from daemon: Error response from daemon: network ub_london has active endpoints
```

**Comment by lavalamp-** [comment]
+1 - this is a pretty big bug, would be great to get some feedback on it

**Comment by caneraydinbey** [comment]
> > version: '2'
> services:
> 
>     couchbase-db:
>         image: couchbase/server
>         volumes:
>             - ./config/cbdata:/opt/couchbase/var
>         ulimits:
>             nofile: 40960
>             core: 10000000
>             memlock: 10000000
>         ports:
>             - '8091-8093:8091-8093'
>             - '11210:11210'
> 
> 


i am having too.

**Comment by vingrad** [comment]
Please fix, I have this bug with Docker 1.12 too.

**Comment by tgeorgiev** [comment]
I am have been experiencing the same with **v1.12.1** and I suspect that the problem was the etcd KV store, as I have been reproducing it quite often with **coreos/etcd:v2.0.8**.
I cannot reproduce it with **coreos/etcd:v2.3.7**, but in any case I am providing my setup and the steps I had to do in order to reproduce it almost every time (v2.0.8)

2 identical nodes 
```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   f1e1b83
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   f1e1b83
 Built:
 OS/Arch:      linux/amd64
```
both are connected to an **coreos/etcd:v2.0.8** discovery service.

Steps to reproduce:
1. Create an overlay network on one of the nodes
2. Create multiple containers on both hosts, connected to this network
3. Delete the containers in parallel
4. Try to delete the network
5. You would get `Error response from daemon: network test-net has active endpoints`
Although inspect would return something like

```
[
    {
        "Name": "test-net",
        "Id": "fd773ef7d7639cb736b99111f675173ced5a7f3b1401ee8d757266b07549709c",
        "Scope": "global",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "10.0.26.0/24",
                    "Gateway": "10.0.26.1/24"
                }
            ]
        },
        "Internal": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
```

Shell script to reproduce: https://gist.github.com/tgeorgiev/b18a3e0d02b0efec09a7341f6d12ec16

**Comment by aboch** [comment]
To recover form this scenario, please follow the steps provided in https://github.com/docker/docker/issues/19261#issuecomment-201085920

**Comment by bhuisgen** [comment]
Same problem here with 1.12 after I restart some hosts

**Comment by danielsand** [comment]
Docker version 1.12.3

Overlay network config:
`[
    {
        "Name": "dev_subnet-01",
        "Id": "a567bd414f80df5668c0aa620ae93f9073df36fdf29adb709486e078e21f0e53",
        "Scope": "global",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.105.0/24",
                    "Gateway": "10.0.105.1/24"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "0bfc7095d5079c61230e2fda8a01702a3930870cb2b57e204a3d2774399e855f": {
                "Name": "xxx",
                "EndpointID": "9e0171cb984c28fb2c593847f631b5f85628aac09269840f46129dc5241ecc7e",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            },
            "25422ecfc05c222bb42ed6654460935314d3833b319f9b05af87ea63a70b8d1d": {
                "Name": "xxx",
                "EndpointID": "c6d69dda7f233fe0e036fce1ad89e8c5fc139a79d42e2b272d2733997af8518a",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            },
            "2a2ef344b45fa550051549856a10c46d69ebf9eb3ec90b4f6174f42624c5241e": {
                "Name": "xxx",
                "EndpointID": "1db7b01fda471f79fd8d82969441f0e97e1e652734394288ea8d775b8b06d940",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            },
            "3e5e9748d3413c250a0baccbe1f2bf15d1aada1e533a462cfbe4f0e2deda397a": {
                "Name": "xxx",
                "EndpointID": "bea926b4e6383f09d6eec44e370f3c79606ea6544918e09715549cae52ea58a5",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            },
            "76081636aa145b0f0edd2284b79a2511ce11a546f1fe561c3d1e9c25e41e1c23": {
                "Name": "xxx",
                "EndpointID": "e2c54edd82d20f10d36226758795cd3510a2ae2f7fbd42672249e1b6e723eb33",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            },
            "8d23d13dba377c0244d583eea42dbc206e212134e5f4162ed7f8d56fdd9a5648": {
                "Name": "xxx",
                "EndpointID": "c1ca18385c2811c0137c4d2a748c526c283a460717d800762db2c80d0fb9b9ac",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            },
            "ep-06ee506ba977b3d1404435e1984a67475fc22b1d6998107ec23da7ab9d255968": {
                "Name": "xxx",
                "EndpointID": "06ee506ba977b3d1404435e1984a67475fc22b1d6998107ec23da7ab9d255968",
                "MacAddress": "02:42:0a:00:69:0d",
                "IPv4Address": "10.0.105.13/24",
                "IPv6Address": ""
            },
            "ep-07fb6955106c36c8a9208221121299d0f7c04639ddb77457d3a2379ea5003b49": {
                "Name": "xxx",
                "EndpointID": "07fb6955106c36c8a9208221121299d0f7c04639ddb77457d3a2379ea5003b49",
                "MacAddress": "02:42:0a:00:69:03",
                "IPv4Address": "10.0.105.3/24",
                "IPv6Address": ""
            },
            "ep-297192d88dd3c28603d64b8475b638f554658d4c0533d47f8a731704cbb89b94": {
                "Name": "xxx",
                "EndpointID": "297192d88dd3c28603d64b8475b638f554658d4c0533d47f8a731704cbb89b94",
                "MacAddress": "02:42:0a:00:69:04",
                "IPv4Address": "10.0.105.4/24",
                "IPv6Address": ""
            },
            "ep-29a81ad4e50fe30031c7abf9db3354f3a71811bc9c7198acfe4602c04bcd09e1": {
                "Name": "xxx",
                "EndpointID": "29a81ad4e50fe30031c7abf9db3354f3a71811bc9c7198acfe4602c04bcd09e1",
                "MacAddress": "02:42:0a:00:69:05",
                "IPv4Address": "10.0.105.5/24",
                "IPv6Address": ""
            },
            "ep-2f5e030eefa056991de51997da8e9d67da7b5593fd71ba840bcf09d9fa202c6a": {
                "Name": "xxx",
                "EndpointID": "2f5e030eefa056991de51997da8e9d67da7b5593fd71ba840bcf09d9fa202c6a",
                "MacAddress": "02:42:0a:00:69:12",
                "IPv4Address": "10.0.105.18/24",
                "IPv6Address": ""
            },
            "ep-3b5591abd1e1f82bd62d599fbf20c145a0a852d0f9fb6fca5d0d6772b87779f0": {
                "Name": "xxx",
                "EndpointID": "3b5591abd1e1f82bd62d599fbf20c145a0a852d0f9fb6fca5d0d6772b87779f0",
                "MacAddress": "02:42:0a:00:69:0e",
                "IPv4Address": "10.0.105.14/24",
                "IPv6Address": ""
            },
            "ep-4ff92475de844bd563804ed8a2de1a88679dcfde83f4715512adffaea4e07835": {
                "Name": "xxx",
                "EndpointID": "4ff92475de844bd563804ed8a2de1a88679dcfde83f4715512adffaea4e07835",
                "MacAddress": "02:42:0a:00:69:16",
                "IPv4Address": "10.0.105.22/24",
                "IPv6Address": ""
            },
            "ep-7a0c7fbf91c30ee52bf3e1ec1ae2ce9d1638de41f41ca5a97173160a41a5c75c": {
                "Name": "xxx",
                "EndpointID": "7a0c7fbf91c30ee52bf3e1ec1ae2ce9d1638de41f41ca5a97173160a41a5c75c",
                "MacAddress": "02:42:0a:00:69:0f",
                "IPv4Address": "10.0.105.15/24",
                "IPv6Address": ""
            },
            "ep-a749ebedab99ad1c403efec42d3fbebcd0dff8b3c9c14bfd943a4d5706792801": {
                "Name": "xxx",
                "EndpointID": "a749ebedab99ad1c403efec42d3fbebcd0dff8b3c9c14bfd943a4d5706792801",
                "MacAddress": "02:42:0a:00:69:15",
                "IPv4Address": "10.0.105.21/24",
                "IPv6Address": ""
            },
            "ep-a87f5c4f093ae788281537178fdb08cdc5832a0ef5a5c8037b5ebe38c4fc1c2b": {
                "Name": "xxx",
                "EndpointID": "a87f5c4f093ae788281537178fdb08cdc5832a0ef5a5c8037b5ebe38c4fc1c2b",
                "MacAddress": "02:42:0a:00:69:0c",
                "IPv4Address": "10.0.105.12/24",
                "IPv6Address": ""
            },
            "ep-c403a9c04f9689119eceb54509b563bf8ed88f2d6a7983d7a03f9fee9c479796": {
                "Name": "xxx",
                "EndpointID": "c403a9c04f9689119eceb54509b563bf8ed88f2d6a7983d7a03f9fee9c479796",
                "MacAddress": "02:42:0a:00:69:10",
                "IPv4Address": "10.0.105.16/24",
                "IPv6Address": ""
            },
            "ep-e3b33899cd2277f843b8fd5784d154aea4c2c303f7b1f33418691794789ac87e": {
                "Name": "xxx",
                "EndpointID": "e3b33899cd2277f843b8fd5784d154aea4c2c303f7b1f33418691794789ac87e",
                "MacAddress": "02:42:0a:00:69:14",
                "IPv4Address": "10.0.105.20/24",
                "IPv6Address": ""
            },
            "ep-ea5ce449f5d54e8ec2992c7b9e2903827bfa0c3cc48109638b5b9dba94a1c804": {
                "Name": "xxx",
                "EndpointID": "ea5ce449f5d54e8ec2992c7b9e2903827bfa0c3cc48109638b5b9dba94a1c804",
                "MacAddress": "02:42:0a:00:69:13",
                "IPv4Address": "10.0.105.19/24",
                "IPv6Address": ""
            },
            "ep-f6112ea8115559713f93db73414ab7508c5cd22c90ae57636e34ebe4703f0328": {
                "Name": "xxx",
                "EndpointID": "f6112ea8115559713f93db73414ab7508c5cd22c90ae57636e34ebe4703f0328",
                "MacAddress": "02:42:0a:00:69:11",
                "IPv4Address": "10.0.105.17/24",
                "IPv6Address": ""
            },
            "ep-fa260ecee85decd5bb2ac2287904363c03758cc0d96610b1e71bf243550c7ad1": {
                "Name": "xxx",
                "EndpointID": "fa260ecee85decd5bb2ac2287904363c03758cc0d96610b1e71bf243550c7ad1",
                "MacAddress": "02:42:0a:00:69:17",
                "IPv4Address": "10.0.105.23/24",
                "IPv6Address": ""
            },
            "f371ef8250070a22df0b35156efd35235342a3e38033371d99418aa496bb960b": {
                "Name": "xxx",
                "EndpointID": "b17b6ef84c5786db5b82173ecc124454c300cc6a34f3ad7cb8b7ad80d21abc23",
                "MacAddress": "",
                "IPv4Address": "",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]`

so lets remove all the active container from the overlay network:
`root@dev-ucp-host1:~# sudo docker network inspect a567bd414f80 |grep End | awk '{ print $2 }' | sed 's/"//g' | sed 's/,//g' | xargs -n1 sudo docker network disconnect -f dev_subnet-01
Error response from daemon: endpoint 9e0171cb984c28fb2c593847f631b5f85628aac09269840f46129dc5241ecc7e not found
Error response from daemon: endpoint c6d69dda7f233fe0e036fce1ad89e8c5fc139a79d42e2b272d2733997af8518a not found
Error response from daemon: endpoint 1db7b01fda471f79fd8d82969441f0e97e1e652734394288ea8d775b8b06d940 not found
Error response from daemon: endpoint bea926b4e6383f09d6eec44e370f3c79606ea6544918e09715549cae52ea58a5 not found
Error response from daemon: endpoint e2c54edd82d20f10d36226758795cd3510a2ae2f7fbd42672249e1b6e723eb33 not found
Error response from daemon: endpoint c1ca18385c2811c0137c4d2a748c526c283a460717d800762db2c80d0fb9b9ac not found
Error response from daemon: endpoint 06ee506ba977b3d1404435e1984a67475fc22b1d6998107ec23da7ab9d255968 not found
Error response from daemon: endpoint 07fb6955106c36c8a9208221121299d0f7c04639ddb77457d3a2379ea5003b49 not found
Error response from daemon: endpoint 297192d88dd3c28603d64b8475b638f554658d4c0533d47f8a731704cbb89b94 not found
Error response from daemon: endpoint 29a81ad4e50fe30031c7abf9db3354f3a71811bc9c7198acfe4602c04bcd09e1 not found
Error response from daemon: endpoint 2f5e030eefa056991de51997da8e9d67da7b5593fd71ba840bcf09d9fa202c6a not found
Error response from daemon: endpoint 3b5591abd1e1f82bd62d599fbf20c145a0a852d0f9fb6fca5d0d6772b87779f0 not found
Error response from daemon: endpoint 4ff92475de844bd563804ed8a2de1a88679dcfde83f4715512adffaea4e07835 not found
Error response from daemon: endpoint 7a0c7fbf91c30ee52bf3e1ec1ae2ce9d1638de41f41ca5a97173160a41a5c75c not found
Error response from daemon: endpoint a749ebedab99ad1c403efec42d3fbebcd0dff8b3c9c14bfd943a4d5706792801 not found
Error response from daemon: endpoint a87f5c4f093ae788281537178fdb08cdc5832a0ef5a5c8037b5ebe38c4fc1c2b not found
Error response from daemon: endpoint c403a9c04f9689119eceb54509b563bf8ed88f2d6a7983d7a03f9fee9c479796 not found
Error response from daemon: endpoint e3b33899cd2277f843b8fd5784d154aea4c2c303f7b1f33418691794789ac87e not found
Error response from daemon: endpoint ea5ce449f5d54e8ec2992c7b9e2903827bfa0c3cc48109638b5b9dba94a1c804 not found
Error response from daemon: endpoint f6112ea8115559713f93db73414ab7508c5cd22c90ae57636e34ebe4703f0328 not found
Error response from daemon: endpoint fa260ecee85decd5bb2ac2287904363c03758cc0d96610b1e71bf243550c7ad1 not found
Error response from daemon: endpoint b17b6ef84c5786db5b82173ecc124454c300cc6a34f3ad7cb8b7ad80d21abc23 not found`

doesnt work - they're are no active container running. But yeah - docker thinks they are....

this is really painfull - ASFIK i can reinstall the complete docker setup to get rid of this... which is bad.
ah - restart reboot of the whole setup didnt bring any hope / luck

**Comment by lavalamp-** [comment]
It would be great to get any amount of feedback from the Docker devs, as this has been an open issue for over 10 months now.

For those of you that are banging your head against the wall, it appears that Docker networking is managed by the ***swarm discovery*** method of your choosing (for instance, I use Consul). You would never know this without doing some serious digging as none of the documentation indicates this, and it's a bit unintuitive. My only work-around to this issue is to completely re-install the swarm discovery instance, as it appears to be a bug in Consul that doesn't recognize removed endpoints. Once Consul has been re-installed, the issue resolves itself.

This bug also appears to be related to the forced removal of endpoints from a Docker network. Unfortunately for anyone that wants to remove machines from a Docker swarm in software and isn't writing software in the same language as docker-machine, this necessitates the use of the force flag (trying to remove without force prompts the user for approval, which is a good bit more difficult to code around than a simple shell invocation). This may have changed as I haven't had the patience to wrestle with this bug in a while, and library bindings for various languages may have been written in that time.

**Comment by juliangamble** [comment]
Here is my process for dealing with this with docker and consul on Docker 1.11
1. Run docker-compose down to stop everything
2. Note the failure to remove the network endpoint
3. Run docker network inspect <networkname>
4. Get the ID of the network from the results above
5. Go into consul into the kv store - under network/docker
6. Find the corresponding entry key
7. Delete the key
8. Run docker-compose down again.

**Comment by bhuisgen** [comment]
We all known how to remove endpoints from CLI and in the K/V but this is not acceptable for a production system where we can't stop all the services. So I'll appreciate a fix for that and without migrating all our working clusters to the new swarm mode as it's too early for production.

**Comment by cpuguy83** [comment]
ping @mrjana Can you take a look at this again? Seems we still have some problems with this in 1.12

**Comment by schmitzhermes** [comment]
same problem. after a restart of the cluster I can't deploy any services from docker-compose sind there already is a service endpoint with the same name. that hit us really hard.

**Comment by dave-pollock** [comment]
As a work around, I wrote this bash script to remove any orphaned endpoints from all docker networks.

```bash
docker network ls -q | while read networkid; do
        docker network inspect $networkid | jq -r '.[0].Containers | keys[]' | while read endpoint; do
                if [[ $endpoint == "ep-"* ]]
                then
                        container=`docker network inspect $networkid | jq -r '.[0].Containers."'"$endpoint"'".Name'`
                        echo "docker network disconnect -f $networkid $container"
                        docker network disconnect -f $networkid $container
                fi
        done
done
```

**Comment by fitz123** [comment]
docker 1.13.1 - problem is still here. Lost endpoint. Node "things" there'is a container, but it was removed a time ago. So I cannot utilize entrypoint, I cannot re-create the service. It's pretty big one. I have to ether restart the daemon or use new port and network for a service =\

**Comment by KevinTHU** [comment]
@thaJeztah this issue is a cocurrent race bug, like two containers on different host being deleted simultaneously, the endpoint_count will be wrongly decreased by the code mentioned below, this code just check the previous value in the remote (like zookeeper), but for a example, if a network has only two containers, and they were all being deleted at the same time, the different dockerd on the host of these two containers all want to update the endpoint_count value from 2 to 1, and all success (by the code logic belowed), finally the endpoint_count becomes to be 1, however, the number of containers in this network is 0 in fact. the network can not be deleted because endpoint_count is not 0, however it should be.
`	if n.ConfigFrom() != "" {
		if t, err := c.getConfigNetwork(n.ConfigFrom()); err == nil {
			if err := t.getEpCnt().DecEndpointCnt(); err != nil {
				logrus.Warnf("Failed to update reference count for configuration network %q on removal of network %q: %v",
					t.Name(), n.Name(), err)
			}
		} else {
			logrus.Warnf("Could not find configuration network %q during removal of network %q", n.configOnly, n.Name())
		}
	}`

`func (ec *endpointCnt) DecEndpointCnt() error {
	return ec.atomicIncDecEpCnt(false)
}`

**Comment by KevinTHU** [comment]
@thaJeztah we can fix this buy by check whether the remote endpoint_value is 2 for the above example (before decrease value, the previous local endpoint_count value need to be the same as remote), then the remote value can be changed form 2 to be 1, otherwise, the whole operation need to be rollback, and the local endpoint_count need to sync from the remote value, and then to be decreased again, for the above example, local endpoint_count need to sync the value to be 1, if the second container delete action is failed, and then the local value 1 need to be decreased to be 0, and then update to the remote, and success, then the endpoint_count will not be wrong by cocurrent race scenario

**Comment by thaJeztah** [comment]
@fitz123 have you tried disconnecting with `-f`?

```
docker network disconnect -f <network-name> <container-id>
```

**Comment by KevinTHU** [comment]
a pull request has been submitted to libnetwork, I think it can fix the problem here, [https://github.com/docker/libnetwork/pull/1780](url)

> when dockerd was restart, the method cleanupLocalEndpoints() would be called, and this method will call method n.getEpCnt().setCnt(uint64(len(epl))), however, setCnt will force update the endpoint_count value in the remote store, and this may cause endpoint_count not right if cocurrently some dockerd on the different host is starting or stopping containers, and the wrong endpoint_count of some globel network will finally make this network cannot be deleted, because before deleting network, dockerd will check whether the value of endpoint_count is {"Count": 0}, however, if this value is wrong because the criteria mentioned above, the network will not be allowed to delete, and we have to manually change the value to be {"Count": 0}, then, the network can be deteted.

this bug has followed docker form 1.9.0 up to now, I hope it can be fixed. thx~

**Comment by KevinTHU** [comment]
@thaJeztah for update, this may(or absolutely for over 30 containers one host) cause endpoint_cnt is not right in zookeeper also, because shutdown timeout(only 15s waiting) in the code below
`shutdownDaemon(d, 15)`
`func shutdownDaemon(d *daemon.Daemon, timeout time.Duration) {
	ch := make(chan struct{})
	go func() {
		d.Shutdown()
		close(ch)
	}()
	select {
	case <-ch:
		logrus.Debug("Clean shutdown succeeded")
	case <-time.After(timeout * time.Second):
		logrus.Error("Force shutdown daemon")
	}
}`
this code may cause "Force shutdown daemon", and if DOCKER_OPTS is not configed with “--live-restore“, all the docker containers will be stopped, but if dockerd is "Force shutdown daemon", some containers will not be stopped completely, then their  endpoint_cnt will not be decreased in zookeeper, because dockerd has been killed at that time.(endpoint will be deleted in zookeeper, but because decreasing endpoint_cnt is the last step for release network, so high probability remained). we can prove it through the log code below in method `func (daemon *Daemon) Shutdown()`, in logs, some containers has log "container stopped"(deleted completely), some does not, but just have log "stopping"(deleted interrupt)
`	if daemon.containers != nil {
		logrus.Debug("starting clean shutdown of all containers...")
		daemon.containers.ApplyAll(func(c *container.Container) {
			if !c.IsRunning() {
				return
			}
			logrus.Debugf("stopping %s", c.ID)
			if err := daemon.shutdownContainer(c); err != nil {
				logrus.Errorf("Stop container error: %v", err)
				return
			}
			if mountid, err := daemon.layerStore.GetMountID(c.ID); err == nil {
				daemon.cleanupMountsByID(mountid)
			}
			logrus.Debugf("container stopped %s", c.ID)
		})
	}`

**Comment by KevinTHU** [comment]
@thaJeztah sorry, I just copyed the docker 1.12.6 code in the above. but I have seen the code in master, also has this problem, if the stoptime `shutdown-timeout` is not set for a big value. however, set a big value fo  `shutdown-timeout` can make this problem work around.

**Comment by greenseedyo** [comment]
@thaJeztah `docker network disconnect -f <network-id> <container-name>` works for me, thanks!!

**Comment by schmunk42** [comment]
I am seeing this with `docker/swarm 1.2.8` and engines with `17.07-09`.

The solution with `docker network disconnect -f` is very cumbersome in my case, since I'd have to ssh into every node and pick the right networks there (when forcing the disconnect).

Any chance to get a fix, which re-syncronizes the store and hosts?

**Comment by antoinetran** [comment]
@schmunk42 How many instances of Swarm Manager and Cluster store do you use? Which cluster store do you use? What is the frequency of the issue?
We might migrate from Swarm 1.2.6 to 1.2.8 to see if this issue is still present, but you seem to say this is still present.

**Comment by schmunk42** [comment]
@antoinetran 

> How many instances of Swarm Manager and Cluster store do you use? 

One manager, one store. (1-2 worker/app nodes)

> Which cluster store do you use? 

Consul

> What is the frequency of the issue?

This usually happens when a worker/app-node gets lost. 
What happens - as far as I understand it - is that on every node in the swarm there's information about all networks. So i.e. also on a manager node there's information about a stack network which was mainly used on the node which just crashed.
When you want to remove the network one (or more) nodes complain about existing endpoints for the network (although they are not on the node itself), if you want to remove these endpoints (containers) no node has those containers (since the one who has it just crashed) and they complain about non-existing containers.
So you are stuck in a dead-end, where you neither can remove the networks nor the containers :(

> We might migrate from Swarm 1.2.6 to 1.2.8 to see if this issue is still present, but you seem to say this is still present.

I thought the same, but there's no difference between 1.2.6 and 1.2.8 regarding this issue, at least not in our tests.

**Comment by antoinetran** [comment]
> What happens - as far as I understand it - is that on every node in the swarm there's information about all networks. So i.e. also on a manager node there's information about a stack network which was mainly used on the node which just crashed.
More precisely, on the cluster store, and there is a cache in /var/lib/docker/network/files/local-kv.db

**Comment by schmunk42** [comment]
Is there a way to clean this cache on the node? 

In https://github.com/moby/moby/issues/19261#issuecomment-337213345 I was inspecting a network via swarm manager, which returned no endpoints. But directly on the node there were endpoints.

**Comment by antoinetran** [comment]
Stop docker, rm this file, restart.

**Comment by mbrewster** [comment]
I was able to resolve the issue by SSH'ing to the node itself, and using `docker network disconnect -f <network ID> <container name>`. Note that *neither* the `Endpoint ID` or `Container ID` resulted in success, nor did running the command against the swarm.

**Comment by piemesi** [comment]
I have the same problem. 
Tried dufferent ways - but the 
```docker network disconnect -f <network ID> <container name>```
didn't work.

Maybe it will be useful:
* for someone - in some of my oppened terminal i using "root" (sudo su) was cleaning the logs for container (via echo "" > $(docker inspect --format='{{.LogPath}}' webserver) )
* so until i wouldn't have exit this terminal (where i as root clean the logs of containers) docker network disconnect -f <network ID> <container name> - wouldn't work
* when i exit - could go on work with network

Wish will help you

**Comment by RobWillis** [comment]
I hit this same issue with 17.12.0-ce.
I posted in this thread originally (last post in): #23302 
Using network disconnect -f <net> <cont> did NOT work for me.
Docker daemon restart hung.
I was forced to reboot.

I'm happy to work with someone to help debug/diagnose the issue if that would help.

**Comment by antoinetran** [comment]
What are your environment?
OS (+kernel) / docker / swarm / type of swarm (classic/swarm mode) / key-value store version

**Comment by RobWillis** [comment]
This is on 
Ubuntu 16.04.2 / Docker version 17.12.0-ce, build c97c6d6 / swarm mode consisting of just 1 manager node.  Attached docker info output.
[dockerinfo.txt](https://github.com/moby/moby/files/1750242/dockerinfo.txt)

**Comment by antoinetran** [comment]
We reproduced again in CentOs 7.4.1708 / docker-ce-17.12.1-ce / swarm image 1.2.8 / zookeeper HA 3.4.11. Had to use the workaround with zookeeper delete endpoint.

**Comment by RobWillis** [comment]
That is great news - is there a fix scheduled?
We have seen this issue happen about 3x a week.  Really very annoying.

**Comment by surbhibakhtiyar** [comment]
Faced the same issue. Unable to remove docker network. There are no running containers but still I am unable to remove the network. 
```
# docker network rm 54f84fde7d37
Error response from daemon: network CFJzQhVL id 54f84fde7d3732777846993f388eda0527939ae65e7e1c45eea76de710bccfd6 has active endpoints
```
```
# docker network inspect 54f84fde7d37
[
    {
        "Name": "CFJzQhVL",
        "Id": "54f84fde7d3732777846993f388eda0527939ae65e7e1c45eea76de710bccfd6",
        "Created": "2018-03-22T16:07:15.465038819Z",
        "Scope": "global",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.16.173.128/26"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]

```
```
$ sudo docker version
Client:
 Version:      17.10.0-ce
 API version:  1.33
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:04:05 2017
 OS/Arch:      linux/amd64
```
```
# sudo docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 17.10.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 seccomp
  WARNING: You're not using the default seccomp profile
  Profile: /etc/docker/seccomp.json
Kernel Version: 3.10.0-693.17.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 125.7GiB
Name: xxxxxxx
ID: ASCQ:2GJR:XBMZ:GBLK:FJ3D:BZJ7:SSHS:VV5A:KEWN:JCZU:U5ZY:A6TX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true
```

**Comment by devzsolt** [comment]
This same issue is still present in Docker version 18.06.1-ce

**Comment by GensaGames** [comment]
Same bug. `Docker version 18.09.0, build 4d60db4`

**Comment by liudalibj** [comment]
Meet same issue at `Version:      17.12.1-ce`

https://github.com/moby/moby/issues/19261#issuecomment-265728121
worked for me

**Comment by g13013** [comment]
The same problem here, I managed to disconnect the network from the container but docker seems to only list a ghost network

```
Client:
 Debug Mode: false

Server:
 Containers: 20
  Running: 20
  Paused: 0
  Stopped: 0
 Images: 470
 Server Version: 19.03.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: w5qdmic62mtn3rdy6w0k1hxmk
  Is Manager: true
  ClusterID: 1bmlhtdmf1d7haiav8v70xqeg
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 192.168.65.3
  Manager Addresses:
   192.168.65.3:2377
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.9.184-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 5.818GiB
 Name: docker-desktop
 ID: EJTH:SLJ7:OKHM:EGAW:4ZHF:Z6WO:YT6I:SDNW:4WKR:N2EL:5ZZJ:YBU5
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 168
  Goroutines: 244
  System Time: 2019-12-16T13:42:14.745269443Z
  EventsListeners: 3
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
 Registry Mirrors:
 Live Restore Enabled: false
 Product License: Community Engine
```

**Comment by jsanchezba** [comment]
> Meet same issue at `Version: 17.12.1-ce`
> 
> [#19261 (comment)](https://github.com/moby/moby/issues/19261#issuecomment-265728121)
> worked for me
> 
> ```
> docker network disconnect -f $networkid $containerName
> ```

This didn't work for me. If i inspect the network that container is still listed there.

Docker version 19.03.12-ce

**Comment by jessesimpson36** [comment]
To remove the zombie networks,  I had to umount the entries in /var/run/docker/netns that corresponded to the zombie network.  umount the entries with the id  as well as entries starting with `lb_`  and `1-` followed by the id of that zombie network.

Then I deleted the folders in /var/run/docker/containerd/   that corresponded to any containers  that were not disconnected to this network (in my case, the containers could not be disconnected because the containers didn't exist).

Then  `systemctl restart docker` or your distros equivalent.

IMO, this is a pretty dangerous solution, so only use it if you don't have any other options. The only reason I chose to take the risk was because this happened on a machine I didn't mind reinstalling if I messed up, and  it was important enough for me to be able to  access  volumes   whose name relied on the name of the docker stack.  As in, without me doing this, I was not able to access a database that was in a container on this stack because the stack couldn't be brought up.

**Comment by felipecrs** [comment]
I've built a small bash function to handle it for me:

```bash
function remove_docker_network() {
  local -r network="$1"

  docker network inspect "${network}" --format '{{ range $key, $value := .Containers }}{{ printf "%s\n" $key}}{{ end }}' | xargs --no-run-if-empty --max-args 1 --verbose -- docker network disconnect --force "${network}"
  docker network rm "${network}"
}

# Example
remove_docker_network my-network
```

Perhaps it helps other people.
